elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +267,254 @@ private Object fetchRelationship(Environment context) {
         }
     }
 
-    /** stub code **/
-    private Object replaceObjects(Environment context) { return null; }
+    /**
+     * handle UPSERT operation
+     * @param context environment encapsulating graphQL's request environment
+     * @return list of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
+
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
+
+        Set<PersistentResource> upsertedObjects = new HashSet<>();
+        EntityDictionary dictionary = context.requestScope.getDictionary();
+
+        for(Map<String, Object> input : context.data.get()) {
+            if(context.isRoot()) {
+                /* get internal reference idFieldName, 'id' in this case */
+                String idFieldName = dictionary.getIdFieldName(dictionary.getEntityClass(context.field.getName()));
+
+                upsertedObjects.add((PersistentResource) upsertObject(idFieldName, input, context.requestScope,
+                        context.outputType));
+            } else {
+                String idFieldName = dictionary.getIdFieldName(context.parentResource.getResourceClass());
+
+                upsertedObjects.add((PersistentResource) upsertObject(idFieldName, input, dictionary, context.parentResource,
+                        context.field, context.outputType, context.requestScope));
+            }
+        }
+        return upsertedObjects;
+    }
+
+    /**
+     * handle upsert on non root-level entities
+     * @param idFieldName id field name
+     * @param input input data object
+     * @param dictionary entity dictionary
+     * @param parentSource parent {@link PersistentResource} object
+     * @param field graphql-java field type
+     * @param outputType graphql-java output type
+     * @return list/set of {@link PersistentResource} objects
+     */
+    private Object upsertObject(String idFieldName, Map<String, Object> input, EntityDictionary dictionary,
+                                PersistentResource parentSource, Field field, GraphQLType outputType,
+                                RequestScope requestScope) {
+        /* fetch id value fed to 'data' argument */
+        Optional<Map.Entry<String, Object>> id = input.entrySet().stream()
```

Comment:
Umm, that would return an `Optional<Object>` type. To get to `Optional<String>`, I'd still have to check if this value is present if so, extract it out and cast as a `String`. So the if-else block right after this line would still be needed. Is there any other way of simplifying without going through all of this?